src/System.Private.CoreLib/shared/System/IO/StreamWriter.cs

```diff
@@ -195,32 +195,67 @@ protected override void Dispose(bool disposing)
             }
             finally
             {
-                // Dispose of our resources if this StreamWriter is closable. 
-                // Note: Console.Out and other such non closable streamwriters should be left alone 
-                if (!LeaveOpen && _stream != null)
+                CloseStreamFromDispose(disposing);
+            }
+        }
+
+        private void CloseStreamFromDispose(bool disposing)
+        {
+            // Dispose of our resources if this StreamWriter is closable. 
+            if (!LeaveOpen && _stream != null)
+            {
+                try
                 {
-                    try
-                    {
-                        // Attempt to close the stream even if there was an IO error from Flushing.
-                        // Note that Stream.Close() can potentially throw here (may or may not be
-                        // due to the same Flush error). In this case, we still need to ensure 
-                        // cleaning up internal resources, hence the finally block.  
-                        if (disposing)
-                        {
-                            _stream.Close();
-                        }
-                    }
-                    finally
+                    // Attempt to close the stream even if there was an IO error from Flushing.
+                    // Note that Stream.Close() can potentially throw here (may or may not be
+                    // due to the same Flush error). In this case, we still need to ensure 
+                    // cleaning up internal resources, hence the finally block.  
+                    if (disposing)
                     {
-                        _stream = null;
-                        _byteBuffer = null;
-                        _charBuffer = null;
-                        _encoding = null;
-                        _encoder = null;
-                        _charLen = 0;
-                        base.Dispose(disposing);
+                        _stream.Close();
                     }
                 }
+                finally
+                {
+                    _stream = null;
+                    _byteBuffer = null;
+                    _charBuffer = null;
+                    _encoding = null;
+                    _encoder = null;
+                    _charLen = 0;
+                    base.Dispose(disposing);
+                }
+            }
+        }
+
+        public override ValueTask DisposeAsync()
+        {
+            if (GetType() != typeof(StreamWriter))
+            {
+                // Since this is a derived StreamWriter, delegate to whatever logic
+                // the derived implementation already has in Dispose.
+                return new ValueTask(Task.Factory.StartNew(s => ((StreamWriter)s).Dispose(), this,
+                    CancellationToken.None, TaskCreationOptions.DenyChildAttach, TaskScheduler.Default));
+            }
+
+            return DisposeAsyncCore();
```

Comment:
The `try / finally` in `DisposeAsyncCore` suggests exceptions are possible. Won't it violate the async dispose pattern to have exceptions escape here vs. returned in `ValueTask`?

Reply:
Hmm, I'm not quite following you.  No exceptions should escape here: if any occur inside DisposeAsyncCore, they'll be stored into the ValueTask that's returned, which presumably is the behavior we'd want.

I have been debating with myself the larger question about whether exceptions should be valid at all.  In general Dispose isn't supposed to throw, and thus DisposeAsync shouldn't either (or return a ValueTask that might fault with an exception), but in cases where an implementation's Dispose is already throwing, it seems reasonable to keep parity on the DisposeAsync, rather than eating the exceptions in cases where they wouldn't be eaten in Dispose.

Opinion?

Reply:
Ugg ... I missed that `DisposeAsyncCore` is `async`. I only checked this method for it.

Yeah I agree that no exceptions is a great principal but reality doesn't always line up with it.